app/controllers/distributions_controller.rb

```diff
@@ -10,6 +10,7 @@ class DistributionsController < ApplicationController
   before_action :enable_turbo!, only: %i[new show]
   skip_before_action :authenticate_user!, only: %i(calendar)
   skip_before_action :authorize_user, only: %i(calendar)
+  skip_before_action :verify_authenticity_token, only: :validate
```

Comment:
In general it's better not to skip this unless there's a good reason to. If the issue is JavaScript, there is a way to send the CSRF token so this passes: https://bloggie.io/@kinopyo/sending-non-get-requests-with-fetch-javascript-api-in-rails

Reply:
This is a strange one, with the current code that extracts form data, the authenticity_token is indeed being submitted, and validated by Rails successfully, at least, the first time. 

But in the scenario where the first time the form is invalid, then user submits again with same/different validation errors, the correct token is being included, but Rails says its invalid.

I tried also with the technique mentioned in that blog post (using header rather than in post body), but the same issue occurs.

Thought it might be due to JS DOM parsing or Stimulus lifecycle keeping a reference to the old token, but that's not the case, it is submitting the latest token.

Requires further investigation if the token verification needs to be there also for the validate endpoint.

Reply:
Did a little more digging into ActionPack (lib/action_controller/metal/request_forgery_protection.rb), where `verify_authenticity_token` is implemented:

It tries three possible ways to validate the given token:
```ruby
compare_with_global_token(csrf_token) ||
            compare_with_real_token(csrf_token) ||
            valid_per_form_csrf_token?(csrf_token)
```

The very first time the form is submitted (and Stimulus controller intercepts and submits `POST /distributions/validate.json`, `compare_with_global_token(csrf_token)` returns true.

But on all subsequent attempts, `compare_with_global_token(csrf_token)` and `compare_with_real_token(csrf_token)` return false, and so it attempts the last option `valid_per_form_csrf_token?(csrf_token)`.

That last method calculates the `correct_token` based on the request path and method:
```ruby
correct_token = per_form_csrf_token(
            session,
            request.path.chomp("/"),
            request.request_method
          )
```

I put a breakpoint here and compared correct_token when the validate endpoint is submitted vs the create endpoint. Since the `request.path` is different, `/distributions.validate.json` vs `/distributions`, it doesn't match for the validate endpoint.

So I think the token that gets generated in the form upon a render from the first validation error, represents a token for the `create` endpoint, and so it won't be correct if the `validate` endpoint is submitted.

Reply:
Is this any help? [https://github.com/JakeChampion/fetch/issues/424#issuecomment-259070410](https://github.com/JakeChampion/fetch/issues/424#issuecomment-259070410) Also, a comment there notes that in Rails 5, every form has its own token, so the link I mentioned wouldn't work any more.